src/new/wallet.js

```diff
@@ -89,28 +82,33 @@ class HathorWallet extends EventEmitter {
     this.passphrase = passphrase;
     this.pinCode = pinCode;
     this.password = password;
-  }
 
-  /**
-   * Called when loading the history of transactions.
-   * It is called every HTTP Request to get the history of a set of addresses.
-   * Usually, this is called multiple times.
-   * The `historyTransactions` contains all transactions, including the ones from a previous call to this method.
-   *
-   * @param {Object} result {historyTransactions, allTokens, newSharedAddress, newSharedIndex, addressesFound}
-   **/
-  onAddressesLoaded(result) {
-    this.emit('more-addresses-loaded', result);
+    this.storage = null;
+    if (storage) {
+      this.storage = storage;
+    } else {
+      // Settings storage data
+      const store = new MemoryStore();
+      this.storage = new Storage()
+      this.storage.setStore(store);
+    }
+    this.storage.setItem('wallet:server', this.conn.currentServer);
+
+    this.onConnectionChangedState = this.onConnectionChangedState.bind(this);
+    this.handleWebsocketMsg = this.handleWebsocketMsg.bind(this);
   }
 
   /**
    * Called when the connection to the websocket changes.
    * It is also called if the network is down.
+   *
+   * @param {Number} newState Enum of new state after change
    **/
-  onConnectionChange(value) {
-    if (value) {
+  onConnectionChangedState(newState) {
+    if (newState === Connection.CONNECTED) {
+      StorageProxy.setStorage(this.storage);
       this.setState(HathorWallet.SYNCING);
-      wallet.loadAddressHistory(0, GAP_LIMIT).then(() => {
+      wallet.loadAddressHistory(0, GAP_LIMIT, this.conn).then(() => {
```

Comment:
Even though our strategy should work most of the time, there may be race conditions where it fails. As `loadAddressHistory` is async, multiple wallets may mix the storages while it is loading.

Our strategy seems to work perfectly for sync methods, but async methods may randomly fail (sometimes without throwing an exception).

Reply:
Yes, that makes sense. I could receive a storage as parameter in async calls, (like I did with `this.conn`, so I know all the sequence of calls will be executed in the same storage).

What do you think?

Reply:
I think it may solve the problem, but we may need a better solution for the future. Is it the only place with async calls?

Reply:
In the new wallet code yes. 

I've added a piece of code to handle it, can you see if it's good?

My first solution was to add an optional parameter for `setItem` and `getItem` in the storage, however the chain of method calls in `loadAddressHistory` is so big that I had to change many methods. I ended up using the solution in the commit because it's less complex to understand and to refactor, if needed.

Reply:
I think it works because the javascript main loop always wait for methods to finish. So, each method will run using the correct store.